### PR TITLE
Cart Chunk and BWF support

### PR DESCRIPTION
**Pull request by naxxfish**
Hi!

Here's the support for cart/bext chunks in Reader().  I've yet to refactor the Writer class, to cope with extra subchunks, which is a little tricker as we'd need to define how the fields are set (I'm guessing the opts that are passed in the constructor, like those in the `fmt` header).  I've also written in a new test, and an example file containing both of these chunks (CC_0101.wav)

One noteworthy change to the behaviour - if Reader encounters an unknown chunk ID, it won't error as it did before - it will emit an `"unknown"` event with the raw buffer data for that chunk.  I needed this for my application - and not knowing about a chunk isn't technically an error, so it's convenient to have the logical separation between actual errors and things we just don't understand.  

Hopefully, all this will make it easier to support other subchunk types in the future - I have my eye on adding `"iXML"` support too.  

I have added one little utility function into reader.js - but this might be better incorporated into the Reader class itself - although there's no real need to.  

Hope it's OK!

Cheers,

### CODE REVIEW

**Comment by martinmunk** [comment]
Fork it!
I've been missing these changes as well, and no development seems to be going on here.

**Comment by LinusU** [comment]
I've recently been added as a maintainer to this repo, I would be happy to assist getting this in 👍 

As it is now, there seems to be some conflicts that needs to be resolved...

**Comment by naxxfish** [comment]
Cool ! What's required ? Happy to fix.

**Comment by LinusU** [comment]
Rebasing on master should probably give you a few conflicts, just fix those and push up again :)

**Comment by naxxfish** [comment]
All conflicts should now be resolved.

File: README.md [hunk 1]
```diff
@@ -4,7 +4,7 @@ node-wav
 [![Build Status](https://secure.travis-ci.org/TooTallNate/node-wav.svg)](http://travis-ci.org/TooTallNate/node-wav)
 
 This module offers streams to help work with Microsoft WAVE files.
-
+Test
```
Comment (LinusU) [inline README.md]: I guess that this was unintentional...

File: README.md [hunk 2]
```diff
@@ -58,6 +58,14 @@ be output, depending on the `audioFormat` property).
 A `"format"` event gets emitted after the WAV header has been parsed.
 
 ### Writer()
+## EXPERIMENTAL
```
Comment (LinusU) [inline README.md]: Should this really be under the `Writer()`?

File: examples/wavinfo.js [hunk 3]
```diff
@@ -22,12 +22,27 @@ if (filename) {
   input = process.stdin;
 }
 
+
+reader.on('cart', function (format) {
+	console.log(JSON.stringify(format))
```
Comment (LinusU) [inline examples/wavinfo.js]: Please use two space indentation

File: examples/wavinfo.js [hunk 4]
```diff
@@ -22,12 +22,27 @@ if (filename) {
...
+})
+
+reader.on('format', function (format) {
+	console.log(JSON.stringify(format))
+})
+
+reader.pipe(devnull());
 input.pipe(reader);
 
+
```
Comment (LinusU) [inline examples/wavinfo.js]: Please remove this empty line

File: examples/wavinfo.js [hunk 5]
```diff
@@ -48,3 +63,4 @@ reader.once('readable', function () {
   console.log('wavDataSize: %d',      reader.subchunk2Size);
   console.log();
 });
+
```
Comment (LinusU) [inline examples/wavinfo.js]: please remove this empty line

File: examples/wavinfo.js [hunk 6]
```diff
@@ -22,12 +22,27 @@ if (filename) {
...
+reader.pipe(devnull());
 input.pipe(reader);
 
+
 // the "readable" event will always come *after* the "format" event, but by now
 // a few final properties will have been parsed like "subchunk2Size" that we want
 // to print out to simulate the wavinfo(1) command
-reader.once('readable', function () {
+reader.once('end', function () {
```
Comment (LinusU) [inline examples/wavinfo.js]: If you need to change this you should update the comment above as well...

File: lib/formats.js [hunk 7]
```diff
@@ -0,0 +1,107 @@
+var format = {
+	0x0000: "Unknown Format",
```
Comment (LinusU) [inline lib/formats.js]: Please use 2 space indentation

File: lib/reader.js [hunk 8]
```diff
@@ -10,6 +10,7 @@ var Transform = require('readable-stream/transform');
 var debug = require('debug')('wave:reader');
 var inherits = util.inherits;
 var f = util.format;
+var formats = require('./formats')
```
Comment (LinusU) [inline lib/reader.js]: `formats` is already declared in this file

File: lib/reader.js [hunk 9]
```diff
@@ -33,6 +34,15 @@ var formats = {
 
 module.exports = Reader;
 
+
+/**
+Utility function
+*/
+function stripNull (mystring)
+{
```
Comment (LinusU) [inline lib/reader.js]: Please follow the current code style

**Comment by naxxfish** [comment]
Thanks - all those issues should be fixed (2013 me has disappointed 2016 me :) )

File: lib/reader.js [hunk 10]
```diff
@@ -89,34 +98,79 @@ Reader.prototype._onFormat = function (chunk) {
...
-  var subchunk1ID = chunk.toString('ascii');
-  this.chunkId = subchunk1ID;
-  if ('fmt ' === subchunk1ID) {
-    this._bytes(4, this._onSubchunk1Size);
-  } else {
-    this.emit('error', new Error(f('bad "fmt id": expected "fmt ", got %j', subchunk1ID)));
-  }
-};
+Reader.prototype._onSubchunkID = function (chunk)
+{
```
Comment (LinusU) [inline lib/reader.js]: Please move `{` to the same line as `function () ` (multiple places)

File: lib/reader.js [hunk 11]
```diff
@@ -89,34 +98,79 @@ Reader.prototype._onFormat = function (chunk) {
...
+	debug('onSubChunkID:', chunk);
+	var subchunkID = chunk.toString('ascii');
+	this.chunkId = subchunkID;
+	debug('onSubChunkID: detected chunk ID', this.chunkId);
+	switch(subchunkID)
+	{
```
Comment (LinusU) [inline lib/reader.js]: Same here

File: lib/reader.js [hunk 12]
```diff
@@ -89,34 +98,79 @@ Reader.prototype._onFormat = function (chunk) {
...
+			break;
+		default:
+			this._bytes(4, this._onSubchunkUnknownSize);
+			break;
+	}
+}
+
+Reader.prototype._onSubchunkUnknownSize = function (chunk)
+{
+	debug('onSubchunkUnknownSize:',chunk);
```
Comment (LinusU) [inline lib/reader.js]: Still using tabs in a lot of places

File: lib/reader.js [hunk 13]
```diff
@@ -89,34 +98,79 @@ Reader.prototype._onFormat = function (chunk) {
...
-  debug('onSubchunk1: %o', chunk);
+Reader.prototype._onSubchunkFmt = function (chunk) {
+  debug('onSubchunkFmt:', chunk);
+  // TODO: support formats other than PCM?
+  // lets do it!
+  
   this.audioFormat = chunk['readUInt16' + this.endianness](0);
+  debug('onSubhcunkFmt:', formats[this.audioFormat])
+  if (formats[this.audioFormat])
+  {
```
Comment (LinusU) [inline lib/reader.js]: same here

File: lib/reader.js [hunk 14]
```diff
@@ -152,89 +206,179 @@ Reader.prototype._onSubchunk1 = function (chunk) {
...
+	this.cart_ProducerAppID = stripNull(chunk.toString('ascii',offset,offset+=64));
+	this.cart_ProducerAppVersion = stripNull(chunk.toString('ascii',offset,offset+=64));
+	this.cart_UserDef = stripNull(chunk.toString('ascii',offset,offset+=64));
+	this.cart_LevelReference = chunk['readUInt32' + this.endianness](offset, offset+=4);
+	this.cart_timers = [];
+	for (var i = 0; i < 8; i++)
+	{ // 8 "post" cart timers
+		this.cart_timers.unshift({ 
+			usage: stripNull(chunk.toString('ascii', offset, offset+=4) ), 
+			value : chunk['readUInt32' + this.endianness](offset, offset+=4)
```
Comment (LinusU) [inline lib/reader.js]: no space before `:` please

File: lib/reader.js [hunk 15]
```diff
@@ -152,89 +206,179 @@ Reader.prototype._onSubchunk1 = function (chunk) {
...
+	this.cart_EndTime = chunk.toString('ascii',offset,offset+=8);
+	this.cart_ProducerAppID = stripNull(chunk.toString('ascii',offset,offset+=64));
+	this.cart_ProducerAppVersion = stripNull(chunk.toString('ascii',offset,offset+=64));
+	this.cart_UserDef = stripNull(chunk.toString('ascii',offset,offset+=64));
+	this.cart_LevelReference = chunk['readUInt32' + this.endianness](offset, offset+=4);
+	this.cart_timers = [];
+	for (var i = 0; i < 8; i++)
+	{ // 8 "post" cart timers
+		this.cart_timers.unshift({ 
+			usage: stripNull(chunk.toString('ascii', offset, offset+=4) ),
```
Comment (LinusU) [inline lib/reader.js]: unbalanced spacing (remove the last space `) )` -> `))`)

File: lib/reader.js [hunk 16]
```diff
@@ -152,89 +206,179 @@ Reader.prototype._onSubchunk1 = function (chunk) {
...
+		end_time : this.cart_EndTime ,
+		producer_app_id : this.cart_ProducerAppID ,
+		producer_app_version : this.cart_ProducerAppVersion ,
+		userdef : this.cart_UserDef,
+		level_reference : this.cart_LevelReference ,
+		timers : this.cart_timers
+	});
+	// carry on!
+	//this._bytes(4, this._onSubChunkID);
+        this._bytes(4, this._onSubchunkID);
```
Comment (LinusU) [inline lib/reader.js]: Wrong indentation

**Comment by LinusU** [comment]
I've added linting to our test so if you rebase on master and run `npm test` you should get all my comments locally :)

**Comment by LinusU** [comment]
I've updated the linting on master again, it would be nice if you could rebase on latest master. As it is now, it seems like the changes on master is showing up in the diff 🤔 a rebase should solve it...

**Pull request closed without merging**